x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/MlPlatformArchitecturesUtil.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.deployment;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoRequestBuilder;
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoResponse;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.monitor.os.OsInfo;
+import org.elasticsearch.plugins.Platforms;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.ml.MachineLearning;
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.core.Strings.format;
+
+public class MlPlatformArchitecturesUtil {
+
+    public static ActionListener<NodesInfoResponse> getArchitecturesSetFromNodesInfoResponseListener(
+        ThreadPool threadPool,
+        ActionListener<Set<String>> architecturesListener
+    ) {
+        return ActionListener.wrap(nodesInfoResponse -> {
+            threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+                architecturesListener.onResponse(getArchitecturesSetFromNodesInfoResponse(nodesInfoResponse));
+            });
+        }, architecturesListener::onFailure);
+    }
+
+    public static NodesInfoRequestBuilder getNodesInfoBuilderWithMlNodeArchitectureInfo(Client client) {
+        return client.admin().cluster().prepareNodesInfo().clear().setNodesIds("ml:true").setOs(true).setPlugins(true);
+    }
+
+    private static Set<String> getArchitecturesSetFromNodesInfoResponse(NodesInfoResponse nodesInfoResponse) {
+        return nodesInfoResponse.getNodes()
+            .stream()
+            .filter(node -> node.getNode().hasRole(DiscoveryNodeRole.ML_ROLE.roleName()))
+            .map(node -> {
+                OsInfo osInfo = node.getInfo(OsInfo.class);
+                return Platforms.platformName(osInfo.getName(), osInfo.getArch());
+            })
+            .collect(Collectors.toUnmodifiableSet());
+    }
+
+    public static void verifyMlNodesAndModelArchitectures(Set<String> architectures, String modelPlatformArchitecture, String modelID)
+        throws IllegalArgumentException, IllegalStateException {
+
+        String architecture = null;
+        Iterator<String> architecturesIterator = architectures.iterator();
+
+        if (architecturesIterator.hasNext()) {
+            architecture = architectures.iterator().next();
+
+            String architecturesStr = architectures.toString();
+            architecturesStr = architecturesStr.substring(1, architecturesStr.length() - 1); // Remove the brackets
```

Comment:
Doing this processing every time the method is called is wasteful, because the string is only needed if there's an error, and we hope there's never an error.

Also, removing the brackets is wasteful as they're added back during logging.

So please delete these two lines.